ShaiRandom/GapShuffler.cs

```diff
@@ -0,0 +1,283 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using ShaiRandom.Generators;
+
+namespace ShaiRandom
+{
+    /// <summary>
+    /// A custom enumerator used to efficiently implement an infinite "gap shuffle"; a shuffle which takes a fixed-size
+    /// set of items and produce a shuffled stream of them such that an element is never chosen in
+    /// quick succession.
+    ///
+    /// Generally, you should use <see cref="EnhancedRandomExtensions.GapShuffle{TItem}(ShaiRandom.Generators.IEnhancedRandom,System.Collections.Generic.IEnumerable{TItem})"/>
+    /// or <see cref="EnhancedRandomExtensions.GapShuffleInPlace{TItem}(ShaiRandom.Generators.IEnhancedRandom,System.Collections.Generic.IList{TItem})"/>
+    /// to get an instance of this, rather than creating one yourself.
+    /// </summary>
+    /// <remarks>
+    /// This enumerator is infinite, so you will need to use it in a foreach loop and break out of it, use something
+    /// like LINQ's .Take(n) function, or use it in a while loop with a condition to break out of it in order to prevent
+    /// an infinite loop.  Until you break out of the loop, it will continue to produce a mostly random shuffled stream
+    /// of the items. These shuffles are spaced so that a single element should always have a large amount of "gap" in
```

Comment:
The way I usually use GapShuffler in Java code is actually not in a loop, but by calling `.next()` when the next single item is needed. Is there an equivalent API here? It might be a default method in one of the enumerating-related interfaces.

Reply:
Similar to how you use MoveNext in an Enumerator by hand:

```CS
var shuffler = rng.GapShuffle(list);
shuffler.MoveNext();
Console.WriteLine(shuffler.Current);
```

It's kind of clunky, but that is how C# defines the IEnumerator interface.  You could of course add a `bool Next(out TItem item)` to make it a bit easier, which might not be a bad idea here.

Reply:
Actually the helper should really just be `void Next()` since the boolean will always be true (these enumerators are infinite).